include/utilities/mdarray/mdarray.hpp

```diff
@@ -0,0 +1,206 @@
+     * @return reference 
+     */
+    const_reference at(const boost::span<const size_t> n) const
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    void insert(const boost::span<const size_t> n, value_type value)
+    {
+        for (size_type i = 0; i < this->m_shape.size(); i++) {
+            if (n[i] >= this->m_shape[i])
+                throw std::out_of_range(
+                    "index " + std::to_string(n[i]) +
+                    " must be less than dimension size " +
+                    std::to_string(this->m_shape[i])
+                );
+        }
+
+        size_type index = this->index(n);
+    
+        if (index >= this->max_size())
+            throw std::out_of_range(
+                "index " + std::to_string(index) +
+                " is larger than maximum possible address index (" +
+                std::to_string(this->max_size()) + ")"
+            );
+        
+        if (index >= this->size() || this->size() == 0)
+            this->m_data.resize(index + 1);
+    
+        this->m_data[index] = value;
+    }
+
+    void insert(std::initializer_list<std::pair<std::initializer_list<size_type>, value_type>> args)
+    {
+        for (const auto& arg : args) {
+            this->insert(arg.first, arg.second);
+        }
+    }
+
+    /**
+     * Allocate this mdarray based on the indices given.
+     * 
+     * @example
+     * mdarray<double> x(3);   // rank 3
+     * x.allocate({5, 10, 10}) // Allocates 5 * 10 * 10 elements
+     * @param n 
+     */
+    void allocate(const boost::span<const size_t> n)
+    {
+        this->m_data.resize(this->index(n));
+    }
+
+    /**
+     * Get the (total) size of this mdarray
+     * (aka the total number of elements).
+     * 
+     * @return size_type
+     */
+    size_type size() const noexcept
+    {
+        return this->m_data.size();
+    }
+
+    size_type max_size() const noexcept
+    {
+        size_type max = 1;
+        for (const auto& v : this->m_shape) {
+            max *= v;
+        }
+        return max;
+    }
+
+    iterator begin() const noexcept {
+        return iterator(*this, 0);
+    }
+
+    iterator end() const noexcept {
+        return iterator(*this, this->m_data.size());
+    }
+
+    /**
+     * Get the rank of this mdarray.
+     * 
+     * @return size_type 
+     */
+    size_type rank() const noexcept
+    {
+        return this->m_shape.size();
+    }
+
+    /**
+     * Get the shape of this mdarray.
+     * 
+     * @return span of the extents per dimension
+     */
+    boost::span<const size_t> shape() const noexcept
+    {
+        return this->m_shape;
+    }
+
+    /**
+     * Index a multi-dimensional set of indices to a single address index.
+     * 
+     * @param n 
+     * @return size_type 
+     */
+    size_type index(const boost::span<const size_t> n) const
+    {
+        size_type index = 0, stride = 1;
+        for (size_type k = 0; k < this->rank(); k++) {
+            index += n[k] * stride;
```

Comment:
Should there be an error check of `n[k] < m_shape[k]`? Could just be `assert` for now, to automatically optimize it out with `-DNDEBUG`